core/validation/failures.py

```diff
@@ -159,9 +192,32 @@ def __str__(self):
             f'Value "{self.value}" is not a valid enum value.'
         )
 
+    def to_user_centered_components(self) -> tuple[str, str, str]:
+        sheet = INTERNAL_TABLE_TO_FORM_TAB[self.sheet]
+        column, section = INTERNAL_COLUMN_TO_FORM_COLUMN_AND_SECTION[self.column]
+        message = (
+            f'For column "{column}", you have entered "{self.value}" which isn\'t correct. You must select an '
+            f"option from the list provided"
+        )
+
+        # additional logic for outcomes to differentiate between footfall and non-footfall
+        if sheet == "Outcome_Data" and self.row_values[4] == "Year-on-year % change in monthly footfall":
+            section = "Footfall Indicator"
+
+        # additional logic for risk location
+        if sheet == "Risk Register":
+            if self.row_values[1]:
+                # project risk
+                section = f"Project {int(self.row_values[1].split('-')[2])} Risks"
+            else:
+                # programme risk
+                section = "Programme Risks"
```

Comment:
I feel like a lot of this could be refactored as it will be reusable across different validation types. Only the message is specific to the error. 

For instance, NonNullableConstraintFailure applies to pretty much every sheet, and so almost all of the code above could go into its UC component. 

I don't think this is urgent or mandatory though, and it's probably better to get the PR in so we can refactor if and when, or put it as a todo.

Reply:
Yep sounds good